command_line/dose_analysis.py

```diff
@@ -0,0 +1,306 @@
+
+    def __init__(self, intensities, dose, params):
+        self.params = params
+        self.stats = None
+        sel = dose >= 0
+        self.intensities = intensities.select(sel)
+        self.dose = dose.select(sel)
+        self.resolution_filter()
+
+    def resolution_filter(self):
+        """Filter arrays on resolution."""
+        if not self.params.d_min and self.params.min_completeness:
+            # Update self.params.d_min using resolution estimate
+            params = phil_defaults.extract().resolutionizer
+            params.nbins = self.params.resolution_bins
+            r = Resolutionizer(self.intensities, params)
+            self.params.d_min = r.resolution_completeness(
+                limit=self.params.min_completeness
+            )
+            logger.info("Estimated d_min: %.2f", self.params.d_min)
+
+        if self.params.d_min or self.params.d_max:
+            sel = flex.bool(self.intensities.size(), True)
+            d_spacings = self.intensities.d_spacings().data()
+            if self.params.d_min:
+                sel &= d_spacings >= self.params.d_min
+            if self.params.d_max:
+                sel &= d_spacings <= self.params.d_max
+            self.intensities = self.intensities.select(sel)
+            self.dose = self.dose.select(sel)
+
+    @classmethod
+    def from_mtz(cls, params, mtz_object):
+        """Initialise the class from mtzf input.
+
+        Args:
+            params: A dose-analysis phil params object
+            mtz_object: An iotbx.mtz.object.
+        """
+        miller_arrays = mtz_object.as_miller_arrays(
+            merge_equivalents=False, anomalous=params.anomalous
+        )
+
+        intensities = None
+        batches = None
+        dose = None
+
+        for ma in miller_arrays:
+            if ma.info().labels == ["BATCH"]:
+                batches = ma
+            elif ma.info().labels == ["DOSE"]:
+                dose = ma
+            elif ma.info().labels == ["I", "SIGI"]:
+                intensities = ma
+            elif ma.info().labels == ["I(+)", "SIGI(+)", "I(-)", "SIGI(-)"]:
+                intensities = ma
+        if not intensities:
+            raise KeyError("Intensity array not found in mtz file")
+        if not batches:
+            raise KeyError("Batch array not found in mtz file")
+
+        indices = mtz_object.extract_original_index_miller_indices()
+        intensities = intensities.customized_copy(indices=indices)
+        batches = batches.customized_copy(indices=indices)
+
+        if params.anomalous:
+            intensities = intensities.as_anomalous_array()
+            batches = batches.as_anomalous_array()
+
+        if dose is None:
+            dose = batches_to_dose(batches.data(), params.dose)
+        else:
+            dose = dose.data()
+
+        return cls(intensities, dose, params)
+
+    @classmethod
+    def from_dials_datafiles(cls, params, experiments, reflection_table):
```

Comment:
I would probably prefer "data_files" rather than "datafiles" here.
```suggestion
    def from_dials_data_files(cls, params, experiments, reflection_table):
```